web/server/codechecker_server/api/mass_store_run.py

```diff
@@ -941,13 +942,54 @@ def get_missing_file_ids(report: Report) -> List[str]:
                 review_status, scc, detection_status, detected_at,
                 run_history_time, analysis_info, analyzer_name, fixed_at)
 
+            if report.annotations:
+                self.__validate_and_add_report_annotations(
+                    session, report_id, report.annotations)
+
             self.__new_report_hashes.add(report.report_hash)
             self.__already_added_report_hashes.add(report_path_hash)
 
             LOG.debug("Storing report done. ID=%d", report_id)
 
         return True
 
+    def __validate_and_add_report_annotations(
+        self,
+        session: DBSession,
+        report_id: int,
+        report_annotation: Dict
+    ):
+        """
+        This function checks the format of the annotations. For example a
+        "timestamp" annotation must be in datetime format. If the format
+        doesn't match then an exception is thrown. In case of proper format the
+        annotation is added to the database.
+        """
+        ALLOWED_ANNOTATIONS = {
+            "timestamp": datetime.fromisoformat,
+            "testsuite": str,
+            "testcase": str
+        }
+
+        for key, value in report_annotation.items():
+            try:
+                ALLOWED_ANNOTATIONS[key](value)
+                session.add(ReportAnnotations(report_id, key, value))
+            except KeyError:
+                raise codechecker_api_shared.ttypes.RequestFailed(
+                    codechecker_api_shared.ttypes.ErrorCode.REPORT_FORMAT,
+                    f"'{key}' is not an allowed report annotation.",
+                    ALLOWED_ANNOTATIONS.keys())
+            except ValueError:
+                raise codechecker_api_shared.ttypes.RequestFailed(
+                    codechecker_api_shared.ttypes.ErrorCode.REPORT_FORMAT,
+                    f"'{value}' has wrong format. '{key}' annotations are "
+                    f"interpreted by "
+                    f"'{ALLOWED_ANNOTATIONS[key].__qualname__}'."
```

Comment:
Can we do something more friendly? I feel like this validation sacrifices the intelligibility of the error message for being somewhat more general, and it feels like a bad trade.

Reply:
I still think this should emit a friendlier error message, even if it means some less elegant code.